File: ch9/SparseTable.cpp
#include <bits/stdc++.h>
using namespace std;

typedef vector<int> vi;

class SparseTable {                              // OOP style
private:
  vi A, P2, L2;
  vector<vi> SpT;                                // the Sparse Table
public:
  SparseTable() {}                               // default constructor

  SparseTable(vi &initialA) {                    // pre-processing routine
    A = initialA;
    int n = (int)A.size();
    int L2_n = (int)log2(n)+1;
    P2.assign(L2_n, 0);
    L2.assign(1<<L2_n, 0);
    for (int i = 0; i <= L2_n; ++i) {
      P2[i] = (1<<i);                            // to speed up 2^i
      L2[(1<<i)] = i;                            // to speed up log_2(i)
    }
    for (int i = 2; i < P2[L2_n]; ++i)
      if (L2[i] == 0)
        L2[i] = L2[i-1];                         // to fill in the blanks

    // the initialization phase
    SpT = vector<vi>(L2[n]+1, vi(n));
    for (int j = 0; j < n; ++j)
      SpT[0][j] = j;                             // RMQ of sub array [j..j]

    // the two nested loops below have overall time complexity = O(n log n)
    for (int i = 1; P2[i] <= n; ++i)             // for all i s.t. 2^i <= n
      for (int j = 0; j+P2[i]-1 < n; ++j) {      // for all valid j
        int x = SpT[i-1][j];                     // [j..j+2^(i-1)-1]
        int y = SpT[i-1][j+P2[i-1]];             // [j+2^(i-1)..j+2^i-1]
        SpT[i][j] = A[x] <= A[y] ? x : y;
      }
  }

  int RMQ(int i, int j) {
    int k = L2[j-i+1];                           // 2^k <= (j-i+1)
    int x = SpT[k][i];                           // covers [i..i+2^k-1]
    int y = SpT[k][j-P2[k]+1];                   // covers [j-2^k+1..j]
    return A[x] <= A[y] ? x : y;
  }
};

int main() {
  // same example as in Chapter 2: Segment Tree
  vi A = {18, 17, 13, 19, 15, 11, 20};
  SparseTable SpT(A);
  int n = (int)A.size();
  for (int i = 0; i < n; ++i)
    for (int j = i; j < n; ++j)
      printf("RMQ(%d, %d) = %d\n", i, j, SpT.RMQ(i, j));
  return 0;
}

File: ch9/SparseTable.py
import math

class SparseTable:
  def __init__(self, A=[]):
    self.A = A
    self.P2 = []
    self.L2 = []
    self.SpT = []

    n = len(self.A)
    L2_n = int(math.log2(n)) + 1
    self.P2 = [0] * (L2_n+1)
    self.L2 = [0] * (2**L2_n+1)
    for i in range(L2_n+1):
      self.P2[i] = 2**i
      self.L2[2**i] = i
    for i in range(2, self.P2[L2_n]):
      if self.L2[i] == 0:
        self.L2[i] = self.L2[i-1]

    self.SpT = [[None] * n for _ in range(self.L2[n]+1)]
    for j in range(n):
      self.SpT[0][j] = j

    for i in range(1, L2_n+1):
      for j in range(n-self.P2[i]+1):
        x = self.SpT[i-1][j]
        y = self.SpT[i-1][j+self.P2[i-1]]
        self.SpT[i][j] = x if A[x] <= A[y] else y

  def RMQ(self, i, j):
    k = self.L2[j-i+1]
    x = self.SpT[k][i]
    y = self.SpT[k][j-self.P2[k]+1]
    return x if self.A[x] <= self.A[y] else y


def main():
  A = [18, 17, 13, 19, 15, 11, 20]
  SpT = SparseTable(A)
  n = len(A)
  for i in range(n):
    for j in range(i, n):
      print('RMQ(%d, %d) = %d' % (i, j, SpT.RMQ(i,j)))


main()
